JS worker import as URL broken with Vite 2.6.0-beta

### Describe the bug

I noticed this while working with audio encoders worker files, Vite fails to resolve JS modules imported as URLs with `?url` suffix, like when importing a CSS worklet file.

It shows this error message while running `vite`:

```
node_modules/vite/dist/node/chunks/dep-b2688e23.js:44099:43: error: [plugin: vite:dep-scan] ENOENT: no such file or directory, lstat '/Users/logaretm/code/vite-project/node_modules/extra-scalloped-border/worklet.js?url'
    44099 │         return normalizePath$4(fs__default.realpathSync(resolved));
          ╵      
```

This only happens if the asset resides in `node_modules`.

This was broken since Vite 2.6.0-beta.0.

I'm not sure but I think this is caused by `getRealPath` introduced in https://github.com/vitejs/vite/pull/4708

Adding a test condition with `queryRE` seems to fix it:

```js
function getRealPath(resolved, preserveSymlinks) {
    if (!preserveSymlinks && browserExternalId !== resolved && !queryRE.test(resolved)) {
        return normalizePath$4(fs__default.realpathSync(resolved));
    }
    return resolved;
}
```

I can submit a PR with this if this is indeed the proper fix for it.

### Reproduction

1. Create a new vite project with `yarn create vite`
2. Upgrade to `vite@2.6.0-beta.2`
3. Add a npm module with worklet file like: `extra-scalloped-border`.
```
yarn add extra-scalloped-border
```

4. Import the worklet file in `main.js`

```js
import workletURL from 'extra-scalloped-border/worklet.js?url'
CSS.paintWorklet.addModule(workletURL)
```

5. Run `yarn vite` and notice the error log


Check the repo with minimal reproduction here:

https://github.com/logaretm/vite-2.6-beta-asset-url-bug

### System Info

```shell
System:
    OS: macOS 11.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 156.22 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.4.1 - /usr/local/bin/node
    Yarn: 1.22.5 - ~/.yarn/bin/yarn
    npm: 7.18.1 - /usr/local/bin/npm
  Browsers:
    Chrome: 95.0.4638.17
    Firefox Developer Edition: 91.0
    Safari: 15.0
    Safari Technology Preview: 15.4
  npmPackages:
    @vitejs/plugin-vue: ^1.9.0 => 1.9.0 
    vite: ^2.6.0-beta.2 => 2.6.0-beta.2
```


### Used Package Manager

yarn

### Logs

```shell
vite:config no config file found. +0ms
  vite:config using resolved config: {
  vite:config   root: '/Users/logaretm/code/vite-project',
  vite:config   base: '/',
  vite:config   mode: 'development',
  vite:config   configFile: undefined,
  vite:config   logLevel: undefined,
  vite:config   clearScreen: undefined,
  vite:config   server: { fs: { strict: undefined, allow: [Array] } },
  vite:config   configFileDependencies: [],
  vite:config   inlineConfig: {
  vite:config     root: undefined,
  vite:config     base: undefined,
  vite:config     mode: undefined,
  vite:config     configFile: undefined,
  vite:config     logLevel: undefined,
  vite:config     clearScreen: undefined,
  vite:config     server: { fs: [Object] }
  vite:config   },
  vite:config   resolve: { dedupe: undefined, alias: [ [Object], [Object] ] },
  vite:config   publicDir: '/Users/logaretm/code/vite-project/public',
  vite:config   cacheDir: '/Users/logaretm/code/vite-project/node_modules/.vite',
  vite:config   command: 'serve',
  vite:config   isProduction: false,
  vite:config   plugins: [
  vite:config     'vite:pre-alias',
  vite:config     'alias',
  vite:config     'vite:modulepreload-polyfill',
  vite:config     'vite:resolve',
  vite:config     'vite:html',
  vite:config     'vite:css',
  vite:config     'vite:esbuild',
  vite:config     'vite:json',
  vite:config     'vite:wasm',
  vite:config     'vite:worker',
  vite:config     'vite:asset',
  vite:config     'vite:define',
  vite:config     'vite:css-post',
  vite:config     'vite:client-inject',
  vite:config     'vite:import-analysis'
  vite:config   ],
  vite:config   build: {
  vite:config     target: [ 'es2019', 'edge88', 'firefox78', 'chrome87', 'safari13.1' ],
  vite:config     polyfillModulePreload: true,
  vite:config     outDir: 'dist',
  vite:config     assetsDir: 'assets',
  vite:config     assetsInlineLimit: 4096,
  vite:config     cssCodeSplit: true,
  vite:config     sourcemap: false,
  vite:config     rollupOptions: {},
  vite:config     commonjsOptions: { include: [Array], extensions: [Array] },
  vite:config     dynamicImportVarsOptions: { warnOnError: true, exclude: [Array] },
  vite:config     minify: 'esbuild',
  vite:config     terserOptions: {},
  vite:config     write: true,
  vite:config     emptyOutDir: null,
  vite:config     manifest: false,
  vite:config     lib: false,
  vite:config     ssr: false,
  vite:config     ssrManifest: false,
  vite:config     reportCompressedSize: true,
  vite:config     chunkSizeWarningLimit: 500,
  vite:config     watch: null
  vite:config   },
  vite:config   env: { BASE_URL: '/', MODE: 'development', DEV: true, PROD: false },
  vite:config   assetsInclude: [Function: assetsInclude],
  vite:config   logger: {
  vite:config     hasWarned: false,
  vite:config     info: [Function: info],
  vite:config     warn: [Function: warn],
  vite:config     warnOnce: [Function: warnOnce],
  vite:config     error: [Function: error],
  vite:config     clearScreen: [Function: clearScreen],
  vite:config     hasErrorLogged: [Function: hasErrorLogged]
  vite:config   },
  vite:config   createResolver: [Function: createResolver],
  vite:config   optimizeDeps: {
  vite:config     esbuildOptions: { keepNames: undefined, preserveSymlinks: undefined }
  vite:config   }
  vite:config } +5ms
  vite:deps Hash is consistent. Skipping. Use --force to override. +0ms

  vite v2.6.0-beta.2 dev server running at:

  > Local: http://localhost:3000/
  > Network: use `--host` to expose

  ready in 186ms.

  vite:time 0.64ms / +0ms
  vite:spa-fallback Rewriting GET / to /index.html +0ms
  vite:time 31.05ms /index.html +42ms
  vite:resolve 2.02ms /@vite/client -> /Users/logaretm/code/vite-project/node_modules/vite/dist/client/client.mjs +0ms
  vite:resolve 0.18ms /main.js -> /Users/logaretm/code/vite-project/main.js +4ms
  vite:load 3.11ms [fs] /@vite/client +0ms
  vite:resolve 0.70ms @vite/env -> /Users/logaretm/code/vite-project/node_modules/vite/dist/client/env.mjs +10ms
  vite:resolve 0.13ms /node_modules/vite/dist/client/env.mjs -> /Users/logaretm/code/vite-project/node_modules/vite/dist/client/env.mjs +1ms
  vite:import-analysis 5.45ms [1 imports rewritten] node_modules/vite/dist/client/client.mjs +0ms
  vite:transform 8.48ms /@vite/client +0ms
  vite:time 19.41ms /@vite/client +53ms
  vite:load 12.70ms [fs] /main.js +12ms
  vite:resolve 0.30ms ./style.css -> /Users/logaretm/code/vite-project/style.css +3ms
  vite:resolve 0.12ms /style.css -> /Users/logaretm/code/vite-project/style.css +0ms
6:21:01 PM [vite] Internal server error: ENOENT: no such file or directory, lstat '/Users/logaretm/code/vite-project/node_modules/extra-scalloped-border/worklet.js?url'
  Plugin: vite:import-analysis
  File: /Users/logaretm/code/vite-project/main.js
      at Object.realpathSync (node:fs:2476:7)
      at getRealPath (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:44099:44)
      at tryNodeResolve (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:43767:16)
      at Context.resolveId (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:43623:28)
      at Object.resolveId (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:50509:55)
      at async TransformContext.resolve (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:50304:23)
      at async normalizeUrl (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:76138:34)
      at async TransformContext.transform (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:76279:57)
      at async Object.transform (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:50566:30)
      at async doTransform (/Users/logaretm/code/vite-project/node_modules/vite/dist/node/chunks/dep-b2688e23.js:65650:29)
```


### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/plugins/resolve.ts
+++ b/packages/vite/src/node/plugins/resolve.ts
@@ -382,7 +382,7 @@ function tryResolveFile(
   } catch (e) {}
   if (isReadable) {
     if (!fs.statSync(file).isDirectory()) {
-      return normalizePath(ensureVolumeInPath(file)) + postfix
+      return getRealPath(file, preserveSymlinks) + postfix
     } else if (tryIndex) {
       if (!skipPackageJson) {
         const pkgPath = file + '/package.json'
@@ -396,7 +396,7 @@ function tryResolveFile(
             targetWeb,
             preserveSymlinks
           )
-          return resolved ? getRealPath(resolved, preserveSymlinks) : resolved
+          return resolved
         }
       }
       const index = tryFsResolve(file + '/index', options, preserveSymlinks)
@@ -478,8 +478,6 @@ export function tryNodeResolve(
   if (!resolved) {
     return
   }
-
-  resolved = getRealPath(resolved, preserveSymlinks)
 
   // link id to pkg for browser field mapping check
   idToPkgMap.set(resolved, pkg)
@@ -919,8 +917,9 @@ function equalWithoutSuffix(path: string, key: string, suffix: string) {
 }
 
 function getRealPath(resolved: string, preserveSymlinks?: boolean): string {
+  resolved = ensureVolumeInPath(resolved)
   if (!preserveSymlinks && browserExternalId !== resolved) {
-    return normalizePath(fs.realpathSync(resolved))
+    resolved = fs.realpathSync(resolved)
   }
-  return resolved
+  return normalizePath(resolved)
 }

